sys/key_rev_large_value.js

```diff
@@ -0,0 +1,433 @@
+
+/**
+ * The chunk size to use for data slicing
+ * @type {number}
+ * @const
+ */
+var CHUNK_SIZE = 31000;
+
+/**
+ * The grace_ttl parameter for a revision policy
+ * @type {number}
+ * @const
+ */
+var GRACE_TTL = 86400;
+
+function range(N) {
+    return Array.apply(null, { length: N }).map(Number.call, Number);
+}
+
+function ChunkedBucket(options) {
+    this.log = options.log || function() {};
+}
+
+ChunkedBucket.prototype._sliceData = function(data) {
+    var slicer;
+    if (Buffer.isBuffer(data)) {
+        slicer = Buffer.prototype.slice;
+    } else {
+        data = data.toString();
+        slicer = String.prototype.substring;
+    }
+    var result = [];
+    for (var index = 0; index < data.length; index += CHUNK_SIZE) {
+        result.push(slicer.call(data, index, index + CHUNK_SIZE));
+    }
+    return result;
+};
+
+ChunkedBucket.prototype._joinData = function(dataParts) {
+    if (!dataParts || !dataParts.length) {
+        return '';
+    } else if (Buffer.isBuffer(dataParts[0])) {
+        return Buffer.concat(dataParts);
+    } else {
+        return ''.concat(dataParts);
+    }
+};
+
+ChunkedBucket.prototype._metaTableName = function(req) {
+    return req.params.bucket + '_meta';
+};
+
+ChunkedBucket.prototype._chunksTableName = function(req) {
+    return req.params.bucket + '_chunks';
+};
+
+ChunkedBucket.prototype._makeMetaSchema = function(opts) {
+    var schemaVersionMajor = 1;
+
+    return {
+        version: schemaVersionMajor * 1000 + (opts.version || 0),
+        options: {
+            compression: opts.compression || [
+                {
+                    algorithm: 'deflate',
+                    block_size: 128
+                }
+            ]
+        },
+        attributes: {
+            key: opts.keyType || 'string',
+            rev: 'int',
+            tid: 'timeuuid',
+            num: 'int',
+            headers: 'json'
+        },
+        index: [
+            { attribute: 'key', type: 'hash' },
+            { attribute: 'rev', type: 'range', order: 'desc' },
+            { attribute: 'tid', type: 'range', order: 'desc' }
+        ]
+    };
+};
+
+ChunkedBucket.prototype._makeChunksSchema = function(opts) {
+    var schemaVersionMajor = 1;
+
+    return {
+        version: schemaVersionMajor * 1000 + (opts.version || 0),
+        options: {
+            compression: opts.compression || [
+                {
+                    algorithm: 'deflate',
+                    block_size: 1024
+                }
+            ]
+        },
+        attributes: {
+            key: opts.keyType || 'string',
+            chunk_id: 'int',
+            tid: 'timeuuid',
+            value: opts.valueType || 'blob'
+        },
+        index: [
+            { attribute: 'key', type: 'hash' },
+            { attribute: 'chunk_id', type: 'hash' },
+            { attribute: 'tid', type: 'range', order: 'desc' }
+        ]
+    };
+};
+
+ChunkedBucket.prototype.createBucket = function(restbase, req) {
```

Comment:
It might be nice to check that only compatible / supported key_rev_value options are passed in, so that code relying on others fails quickly.